### DISCUSSION TITLE
Client side locale

### DISCUSSION THREAD

**Discussion by gpo2307** [primefaces/community#2735] [PrimeFaces] [answered]
Hi !

We are migrating our software with Primefaces from version 13 to 14. According to primefaces migration guide, for AutoComplete, emptyMessage property have been moved to client side locale. So i tried to put this property in client side (we already have a js file) put i never see the corresponding translation. I tried everything (see on bottom my example). Thank you by advance

```js
PrimeFaces.locales ['fr'] = {
                emptyMessage: 'myTest994',
                "primefaces.autoComplete.emptyMessage": "myTest9",
                "primefaces.AutoComplete.emptyMessage": "myTest11",
                "autoComplete.emptyMessage": "myTest12",
                "AutoComplete.emptyMessage": "myTest13",
                messages: {
                    "primefaces.autoComplete.emptyMessage": "myTest",
                    "primefaces.AutoComplete.emptyMessage": "myTest2",
                    "autoComplete.emptyMessage": "myTest3",
                    "AutoComplete.emptyMessage": "myTest4",
                    emptyMessage: 'myTest995',
                },
                aria: {
                    "primefaces.autoComplete.emptyMessage": "myTest5",
                    "primefaces.AutoComplete.emptyMessage": "myTest7",
                    "autoComplete.emptyMessage": "myTest6",
                    "AutoComplete.emptyMessage": "myTest8",
                    emptyMessage: 'myTest996',
                }
             };

// custom PF labels
PrimeFaces.locales["fr"] = $.extend(true, {}, PrimeFaces.locales["fr"], {
    emptyMessage: 'myTest997',
    "primefaces.autoComplete.emptyMessage": "myTest14",
    "primefaces.AutoComplete.emptyMessage": "myTest15",
    "autoComplete.emptyMessage": "myTest16",
    "AutoComplete.emptyMessage": "myTest17",
    aria: {
         "primefaces.autoComplete.emptyMessage": "myTest18",
        "primefaces.AutoComplete.emptyMessage": "myTest19",
        "autoComplete.emptyMessage": "myTest20",
        "AutoComplete.emptyMessage": "myTest21",
        emptyMessage: 'myTest998',
    },
    messages: {
        // components
        "primefaces.autoComplete.emptyMessage": "myTest22",
        "primefaces.AutoComplete.emptyMessage": "myTest23",
        "autoComplete.emptyMessage": "myTest24",
        "AutoComplete.emptyMessage": "myTest25",
        emptyMessage: 'myTest999',
    },
});
```

**Comment by melloware** [comment, marked as answer]
you can just do...

```js
PrimeFaces.locales["fr"].emptyMessage = "Your Empty Message";
```

The key is this file MUST load AFTER the PrimeFaces locale loads so check your F12 network history and see if your JS is being loaded before `locale-fr.js` is?

you can also do this.

```xml
<p:autoComplete completeMethod="#{frag.actionAutocomplete}"  widgetVar="autoComplete" value="#{frag.selectedValue}" > 
   <f:attribute name="widgetPostConstruct" value="PF('autoComplete').emptyMessage = '#{frag.myCustomEmptyMessage}';"/>
</p:autoComplete>
```

**Reply by gpo2307** [reply]
Thanks for you answer ! 

I tried :

PrimeFaces.locales["fr"].emptyMessage = "Your Empty Message";

But it's always the same

I've checked in my browser and locale-fr.js is loaded before mine (primefaces.locale.js)

<img width="384" alt="image" src="https://github.com/user-attachments/assets/c52d277d-8074-4d1a-a952-c8156f497d4d">


The third manner works but it would be much better if i could use locales in js.

Do you have an idea what could be wrong ?

**Reply by melloware** [reply]
Ahhh I see I created a little test can you can run with `mvn clean jetty:run`.

[discussions-2735.zip](https://github.com/user-attachments/files/17113734/discussions-2735.zip)

you need to set `emptySearchMessage` not `emptyMessage`

```js
PrimeFaces.locales["fr"].emptyMessage = "Your Empty Message";
PrimeFaces.locales["fr"].emptySearchMessage = "Your Empty Search Message";
```

**Comment by gpo2307** [comment]
Thank you, it worked wity emptySearchMessage. But how can i determine the good key to set in my js file ? Because controls properties aren't the same according to primefaces migration guide . (https://primefaces.github.io/primefaces/13_0_0/#/../migrationguide/14_0_0)

**Reply by melloware** [reply]
good point I just commited a fix updating the migration guide: https://github.com/primefaces/primefaces/blob/master/docs/migrationguide/14_0_0.md#autocomplete

## AutoComplete

  * `dropdownAriaLabel` moved to client side locale.js key `choose`
  * `emptyMessage` moved to client side locale.js key `emptySearchMessage`
  * `resultsMessage`  moved to client side locale.js key `searchMessage`

**Reply by gpo2307** [reply]
Thank you very much. Can you also tell me what are the keys for  pickList : 

addLabel, addAllLabel, removeLabel, removeAllLabel, moveUpLabel,  moveTopLabel, moveDownLabel,  moveBottomLabel

It is also in this migration guide.

Thank you by advance

**Reply by melloware** [reply]
here: https://github.com/primefaces/primefaces/blob/d7dd16abf9f2f52918220ed304599d4ff445a039/primefaces/src/main/resources/META-INF/resources/primefaces/locales/locale-fr.js#L91-L98

Since those are in ARIA section it will be like this.

```js
PrimeFaces.locales["fr"].aria.moveAllToSource = "Tout déplacer vers la source";
```
